sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -849,12 +850,18 @@ object TypeCoercion {
       case Add(l @ DateType(), r @ IntegerType()) => DateAdd(l, r)
       case Add(l @ IntegerType(), r @ DateType()) => DateAdd(r, l)
       case Subtract(l @ DateType(), r @ IntegerType()) => DateSub(l, r)
-      case Subtract(l @ DateType(), r @ DateType()) => DateDiff(l, r)
-      case Subtract(l @ TimestampType(), r @ TimestampType()) => TimestampDiff(l, r)
+      case Subtract(l @ DateType(), r @ DateType()) =>
+        if (SQLConf.get.getConf(SQLConf.LEGACY_DATES_SUBTRACTION)) {
```

Comment:
Could we do not add `spark.sql.legacy.datesSubtraction.enabled` and use `spark.sql.dialect` here?
`spark.sql.dialect=SPARK` -> `SubtractDates(l, r)`
`spark.sql.dialect=PostgreSQL` -> `DateDiff(l, r)`

Reply:
The `spark.sql.legacy.datesSubtraction.enabled` was added to allow users to restore old behavior. The `spark.sql.dialect=PostgreSQL` is so global settings for this.

Reply:
We could restore the previous behavior when 
```spark.sql.legacy.datesSubtraction.enabled || spark.sql.dialect=PostgreSQL```

Reply:
How about `spark.sql.dialect.spark.datesSubtraction.enabled`

Reply:
We want to switch to the new behavior by default, with a legacy config to restore the behavior. This is different from other ANSI behaviors, which we don't want to switch to by default.

Reply:
OK. Thank you all.